enhancements/cloud-integration/azure/azure-workload-identity.md

```diff
@@ -0,0 +1,370 @@
+- Creating Managed Identity infrastructure with federated credentials for OpenShift operator `ServiceAccounts` (identified by namespace & name) and to output secrets containing the `clientID` of the Managed Identity to be provided as manifests for the installer. This command will process `CredentialsRequest` custom resources to identify service accounts that will be associated with Managed
+  Identities in Azure as federated credentials. For self-managed installation, `CredentialsRequests` will be exracted from the release image.
+
+```sh
+$ ./ccoctl azure
+Creating/updating/deleting cloud credentials objects for Azure
+
+Usage:
+  ccoctl azure [command]
+
+Available Commands:
+  create-all                Create key pair, identity provider and Azure Managed Identities
+  create-identity-provider  Create identity provider
+  create-key-pair           Create a key pair
+  create-managed-identities Create Azure Managed Identities
+  delete                    Delete Azure identity provider and Managed Identity infrastructure
+
+Flags:
+  -h, --help   help for azure
+
+Use "ccoctl azure [command] --help" for more information about a command.
+```
+
+#### Azure CredentialsRequests ServiceAccounts
+
+`CredentialsRequests` for the Azure platform must now list
+[ServiceAccountNames](https://github.com/openshift/cloud-credential-operator/blob/1f7a2602bf8a9ddec5d8fc29f77215697d9e7c07/pkg/apis/cloudcredential/v1/types_credentialsrequest.go#L57-L62)
+in order to for `ccoctl` to be able to create federated credentials
+for an Azure Managed Identity that are associated with the `name` and
+`namespace` of the `ServiceAccount`. Example:
+[aws-ebs-csi-driver-operator](https://github.com/openshift/cluster-storage-operator/blob/f1ddb697afb3c33d6d45936e58fad101abe26f13/manifests/03_credentials_request_aws.yaml#L11-L13).
+
+#### Credentials secret
+
+OpenShift operators currently obtain their long-lived credentials from a config secret with the following format:
+
+```yaml
+apiVersion: v1
+data:
+  azure_client_id: <client id>
+  azure_client_secret: <client secret>
+  azure_region: <region>
+  azure_resource_prefix: <resource group prefix eg. "abutcher-az-t68n4">
+  azure_resourcegroup: <resource group eg. "abutcher-az-t68n4-rg">
+  azure_subscription_id: <subscription id>
+  azure_tenant_id: <tenant id>
+kind: Secret
+type: Opaque
+```
+
+We propose that when utilizing Azure Workload Identity, the credentials secret will contain an `azure_client_id` that is the `clientID` of the Managed Identity provisioned by `ccoctl` for the operator. The `azure_client_secret` key will be absent and instead we can provide the path to the mounted `ServiceAccount` token as an `azure_federated_token_file` key; the path to the mounted token is well
```

Comment:
I was expecting that we will set the secret as described here, those would be reflected in the [cloud conf](https://cloud-provider-azure.sigs.k8s.io/install/configs/) and used by the Azure driver to decide on authentication method. 

But looking at the proposed driver/adal [changes](https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/1193/files#diff-5a2bd0a0d9675c240af38567ca20a741d3e1be44c90d83765ac2615ff8a4f000R935-R945) now, it seems like the WI feature will rely only on having specific env variables set - we won't have the option to put `azure_federated_token_file` key into the config. Also, the env variable presence would normally be taken care of by the webhook which we decided to deploy but not use for storage driver deployments.

In all cases I think it should not be up to a driver operator to decide on enablement/disablement of this feature so I think we have two options:

1) Request Azure patches so that the cloud config will be able to hold new values for WI (client id, tenant id, federated token path). Then the driver/adal lib will decide on the authentication based on that config. That config would be set in scope of CCO.

2) Patch CCO so that it will provide secrets to driver operator, if the secrets are present driver operator would load the env variables for the driver to enable WI (`AZURE_CLIENT_ID`,  `AZURE_TENANT_ID`, `AZURE_FEDERATED_TOKEN_FILE` )

I think option 1) is more clean, consistent and matches initial plan. Also we could use the [credentials injector](https://github.com/openshift/azure-file-csi-driver-operator/blob/6120df818b59919133f5b686238d46a1c866c724/assets/node.yaml#L33-L52) that can update the config file, if we ever need to do it.

@abutcher Maybe I missed something, what approach would you prefer?

Reply:
@RomanBednar Yes, I think option 1 would be preferred and would match the existing configuration methods.

Reply:
Do we need to do something to make the enhancement more clear on this point?

Reply:
@RomanBednar `ccoctl` will output secrets with the managed identity information in either scenario. The difference will come from how we pass the client information in that secret on to operators either as environment variables specified in deployments or in the config secret referenced by deployments.